app/src/main/java/br/com/dcarv/arcmovies/presentation/view/MainActivity.kt

```diff
@@ -0,0 +1,293 @@
+import kotlinx.android.synthetic.main.content_main.*
+import okhttp3.OkHttpClient
+import okhttp3.logging.HttpLoggingInterceptor
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+
+/**
+ * An implementation of IMainActivity for Android.
+ * Shows the list of upcoming movies
+ *
+ * @author Danilo Carvalho
+ */
+class MainActivity : AppCompatActivity(), IMainView,
+    SearchView.OnQueryTextListener, RegionSettingsPickerDialogFragment.Listener {
+    // Presenter
+    private lateinit var presenter: IMainPresenter
+
+    // Network Dependencies
+    private lateinit var service: TMDBService
+    private lateinit var moviesProvider: IUpcomingMoviesListProvider
+    private lateinit var genreProvider: IGenreProvider
+    private lateinit var countriesProvider: ICountriesListProvider
+    private lateinit var languagesProvider: ILanguagesListProvider
+    private lateinit var transtationsProvider: ITranslationsProvider
+    private lateinit var preferencesProvider: IPreferencesProvider
+
+    // Movies Recycler Adapter
+    lateinit var adapter: MoviesAdapter
+
+    private var searchView: SearchView? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        setSupportActionBar(toolbar)
+
+        // TODO: use Dependency Injection to inject these dependencies instead
+        // retrofit logging
+        val client = if (BuildConfig.DEBUG) {
+            // retrofit logging
+            val loggingInterceptor = HttpLoggingInterceptor()
+            loggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
+
+            OkHttpClient.Builder()
+                .addInterceptor(loggingInterceptor)
+                .build()
+        } else {
+            OkHttpClient.Builder().build()
+        }
+        val retrofit = Retrofit.Builder()
+            .baseUrl(TMDB_URL)
+            .client(client)
+            .addConverterFactory(GsonConverterFactory.create())
+            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
+            .build()
+        service = retrofit.create(TMDBService::class.java)
+        moviesProvider = TMDBUpcomingMoviesListProvider(
+            TMDB_API_KEY,
+            service,
+            mapper = TMDBUpcomingMoviesListMapper
+        )
+        genreProvider = TMDBGenreProvider(TMDB_API_KEY, service, mapper = TMDBGenreMapper)
+        countriesProvider =
+                TMDBCountryListProvider(TMDB_API_KEY, service, mapper = TMDBCountryMapper)
+        languagesProvider =
+                TMDBLanguagesListProvider(TMDB_API_KEY, service, mapper = TMDBLanguageMapper)
+        transtationsProvider = TMDBTranslationsProvider(TMDB_API_KEY, service)
+        preferencesProvider = SharedPreferencesProvider(this.applicationContext)
+
+        // Get previous presenter or create new
+        presenter = lastCustomNonConfigurationInstance as? IMainPresenter ?: MainPresenter(
```

Comment:
Would you mind giving me some details in how this `lastCustomNonConfigurationInstance` works?

Reply:
Sure. When Android destroys an activity because of a configuration change it calls the method "onRetainCustomNonConfigurationInstance()". Whatever this method returns will be retained in memory until the activity is recreated by the OS. Then the activity can call "getLastCustomNonConfigurationInstance()" to retrieve this object.

Reply:
This is super awesome! I'm very surprised I'd never ran into this before. I was so happy with the new ViewModel but little did I know we had this all along...

Reply:
I know, right? I only found out about it a couple of months ago and I was very surprised this existed and no one talked about it.